<a href="https://colab.research.google.com/github/hamsterjoa/Kbank_playstore_review_analysis/blob/master/sentiment_by_LogisticRegression.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import os, sys  
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [16]:
import io
import pandas as pd

from sklearn import metrics
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.linear_model import LogisticRegression
from sklearn.feature_extraction.text import TfidfVectorizer

### 데이터 로드

In [19]:
# 파일 불러오기
test = pd.read_pickle("/content/drive/My Drive/Colab Notebooks/test_data_for_modeling.pkl")
train = pd.read_pickle("/content/drive/My Drive/Colab Notebooks/train_data_for_modeling.pkl")

## 모델링
- 전처리한 main_pos 칼럼을 이용해 모델 학습을 진행합니다.
- train 셋에서 70%를 학습에, 25%를 모델이 유의미한지 검증하는데 사용합니다.

### train/validation 정의

In [20]:
train.shape

(1530, 5)

In [21]:
data = train.copy()
train, val = train_test_split(data)
train.reset_index(inplace=True) # 전처리 과정에서 데이터가 뒤섞이지 않도록 인덱스를 초기화
val.reset_index(inplace=True)

print( 'train 데이터 셋 모양 :', train.shape)
print( 'val 데이터 셋 모양 :', val.shape)

train 데이터 셋 모양 : (1147, 6)
val 데이터 셋 모양 : (383, 6)


### 벡터화

In [25]:
X_train = train.main_pos # training 데이터에서 문서 추출
y_train = train.rating # training 데이터에서 라벨 추출

In [26]:
vectorizer = CountVectorizer()
vectorizer.fit(X_train) # countvectorizer 학습
X_train_vec = vectorizer.transform(X_train)

### 모델 학습

In [27]:
model = LogisticRegression(max_iter=500)
model.fit(X_train_vec, y_train)

LogisticRegression(max_iter=500)

### 검증셋으로 모델 성능 검증

In [28]:
X_val = val.main_pos # validation 데이터에서 전처리된 문서 추출
y_val = val.rating # validation 데이터에서 라벨 추출

X_val_vec = vectorizer.transform(X_val) # train셋으로 fit한 벡터라이저 이용해 transform

In [29]:
y_pred = model.predict(X_val_vec)
print(y_pred)

[1 1 1 3 1 1 1 5 1 1 1 1 5 1 1 5 1 1 1 5 5 5 1 1 1 1 1 1 1 1 1 1 5 1 1 1 1
 1 1 1 1 1 1 1 1 3 1 1 1 1 1 1 1 1 1 1 1 5 1 1 1 1 1 1 1 1 1 3 1 1 1 1 1 1
 5 1 1 5 1 1 1 1 2 1 1 1 1 1 1 1 3 1 1 1 1 3 1 1 3 1 1 5 1 1 1 1 1 1 5 1 1
 1 1 1 1 1 2 1 1 1 1 1 1 1 1 1 1 1 1 3 1 1 1 1 1 1 1 2 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 3 5 1 1 1 1 1 1 1 1 1 1 1 1 5 1 1 1 1 1 1 1 1 1 1 1 5 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 5 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 2 5 1 5 1 1 1 5 1 1 1 1 1 1 1 1 1 1 3 4 1 1 1 1
 1 1 4 5 1 1 1 1 1 1 1 2 1 1 5 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 1 1 1
 1 1 1 3 1 1 1 1 1 1 5 1 1 1 1 1 5 1 1 5 1 1 1 1 1 1 2 5 1 1 1 1 1 1 1 1 1
 1 1 3 1 1 1 1 1 1 1 1 5 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 3 1 1 2 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1]


In [30]:
print('Logistic Regression 의 예측 정확도는', round(metrics.accuracy_score(y_val, y_pred),3)) 

Logistic Regression 의 예측 정확도는 0.76


### TF-IDF

In [31]:
X_train = train.main_pos #training 데이터에서 문서 추출
y_train = train.rating #training 데이터에서 라벨 추출

# TF-IDF Vectorization 적용하여 학습 데이터셋과 테스트 데이터 셋 변환. 
tfidf_vect = TfidfVectorizer()
tfidf_vect.fit(X_train)

X_val = val.main_pos #validation 데이터에서 전처리된 문서 추출
y_val = val.rating #validation 데이터에서 라벨 추출

X_train_tfidf_vect = tfidf_vect.transform(X_train)
X_val_tfidf_vect = tfidf_vect.transform(X_val) # train셋으로 fit한 벡터라이저 이용해 transform
print('학습 & 테스트 데이터 Text의 TfidfVectorizer Shape:',X_train_tfidf_vect.shape, X_val_tfidf_vect.shape)

학습 & 테스트 데이터 Text의 TfidfVectorizer Shape: (1147, 5370) (383, 5370)


In [32]:
# LogisticRegression을 이용하여 학습/예측/평가 수행. 
lr_clf = LogisticRegression(solver='liblinear') 
lr_clf.fit(X_train_tfidf_vect , y_train)
pred = lr_clf.predict(X_val_tfidf_vect)
print('TF-IDF Logistic Regression 의 예측 정확도는 {0:.3f}'.format(metrics.accuracy_score(y_val ,pred)))

TF-IDF Logistic Regression 의 예측 정확도는 0.762


*TF-IDF를 적용하여 검증 셋을 예측해본 결과 0.762%로 정확도가 상승한 것을 알 수 있습니다.*

## test 분류

In [42]:
X_train = data.main_pos 
y_train = data.rating 

# TF-IDF Vectorization 적용하여 학습 데이터셋과 테스트 데이터 셋 변환. 
tfidf_vect = TfidfVectorizer(ngram_range=(1,2), max_df=300)
tfidf_vect.fit(X_train)

X_train_tfidf_vect = tfidf_vect.transform(X_train)

print(X_train_tfidf_vect.shape)

(1530, 30230)


In [43]:
# LogisticRegression을 이용하여 학습/예측/평가 수행. 
lr_clf = LogisticRegression(solver='liblinear', C = 10) 
lr_clf.fit(X_train_tfidf_vect , y_train)

LogisticRegression(C=10, solver='liblinear')

In [55]:
X_test = test.main_pos
y_test = test.rating

X_test_vec = tfidf_vect.transform(X_test)

In [56]:
pred_test = lr_clf.predict(X_test_vec)
print(pred_test)

[1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 5 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 5 1 1 1 1 1 1 5 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 3 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 5 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 5 1 1 1 1 1 1 5 1 1 1 1 1 1 1 1 1 1 1 1 5 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 3 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 3 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 5 1 1 1 1 5 1 1 1 1 1 1 3 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 5 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 5 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 5 1 1 1 1 1 1 1 1 1 1 1 1 1 5 3 1 1 1 1 1 1 1 1 1 1 1 5 1 1 1
 1 1 1 1 1 1 1 1 1 1 5 5 1 1 1 1 1 1 1 1 5 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 5 1 1 1 1 1 1 1 1 1 1 1 1 1 1 5 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 5 1 1 1 5 5 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 

In [57]:
print('Logistic Regression 의 예측 정확도는', round(metrics.accuracy_score(y_test, pred_test),3))

Logistic Regression 의 예측 정확도는 0.69
